src/ripple/protocol/STValidation.h

```diff
@@ -21,89 +21,152 @@
 #define RIPPLE_PROTOCOL_STVALIDATION_H_INCLUDED
 
 #include <ripple/protocol/PublicKey.h>
-#include <ripple/protocol/SecretKey.h>
 #include <ripple/protocol/STObject.h>
+#include <ripple/protocol/SecretKey.h>
 #include <cstdint>
+#include <functional>
 #include <memory>
 
 namespace ripple {
 
 // Validation flags
-const std::uint32_t vfFullyCanonicalSig    = 0x80000000; // signature is fully canonical
+const std::uint32_t vfFullyCanonicalSig =
+    0x80000000;  // signature is fully canonical
 
-class STValidation final
-    : public STObject
-    , public CountedObject <STValidation>
+class STValidation final : public STObject, public CountedObject<STValidation>
 {
 public:
-    static char const* getCountedObjectName () { return "STValidation"; }
+    static char const*
+    getCountedObjectName()
+    {
+        return "STValidation";
+    }
 
     using pointer = std::shared_ptr<STValidation>;
-    using ref     = const std::shared_ptr<STValidation>&;
+    using ref = const std::shared_ptr<STValidation>&;
 
-    enum
-    {
-        kFullFlag = 0x1
-    };
+    enum { kFullFlag = 0x1 };
+
+    /** Construct a STValidation from a peer.
+
+        Construct an STValidation from serialized data previously shared by a
+        peer.
+
+        @param sit Iterator over serialized data
+        @param lookupNodeID Find the Node ID based on the public key used to
+                            sign the validation. For manifest based validators,
+                            this should be the NodeID of the master public key.
+        @param checkSignature Whether to verify the data was signed properly
 
-    // These throw if the object is not valid
-    STValidation (SerialIter & sit, bool checkSignature = true);
+        @note Throws if the object is not valid
+    */
+    STValidation(
+        SerialIter& sit,
+        std::function<NodeID(PublicKey const&)> const& lookupNodeID,
```

Comment:
I don't love `std::function` if it can be avoided (it type erases and may allocate). This function is used to calculate a `NodeID`. One solution: we could make this parameter a lambda and then delegate to a private constructor that takes the `NodeID` directly. That way you don't have to put the much of an implementation in the header file.

Reply:
The `PublicKey` that is the argument to the function requires initializing the `STObject` base class first, so I had trouble delegating to a private constructor. I think putting everything in the header and using a template isn't so bad in this case.